File: pixels-core/src/main/java/cn/edu/ruc/iir/pixels/core/reader/PixelsReaderOption.java
package cn.edu.ruc.iir.pixels.core.reader;

import cn.edu.ruc.iir.pixels.core.PixelsPredicate;

import java.util.Optional;

/**
 * pixels
 *
 * @author guodong
 */
public class PixelsReaderOption
{
    private String[] includedCols = new String[0];
    private PixelsPredicate predicate = null;
    private boolean skipCorruptRecords = false;
    private boolean tolerantSchemaEvolution = true;    // this may lead to column missing due to schema evolution
    private int rgStart = 0;
    private int rgLen = -1;     // -1 means reading to the end of the file

    public PixelsReaderOption()
    {
    }

    public void includeCols(String[] columnNames)
    {
        this.includedCols = columnNames;
    }

    public String[] getIncludedCols()
    {
        return includedCols;
    }

    public void predicate(PixelsPredicate predicate)
    {
        this.predicate = predicate;
    }

    public Optional<PixelsPredicate> getPredicate()
    {
        if (predicate == null)
        {
            return Optional.empty();
        }
        return Optional.of(predicate);
    }

    public void skipCorruptRecords(boolean skipCorruptRecords)
    {
        this.skipCorruptRecords = skipCorruptRecords;
    }

    public boolean isSkipCorruptRecords()
    {
        return skipCorruptRecords;
    }

    public void rgRange(int rgStart, int rgLen)
    {
        this.rgStart = rgStart;
        this.rgLen = rgLen;
    }

    public int getRGStart()
    {
        return this.rgStart;
    }

    public int getRGLen()
    {
        return this.rgLen;
    }

    public void tolerantSchemaEvolution(boolean tolerantSchemaEvolution)
    {
        this.tolerantSchemaEvolution = tolerantSchemaEvolution;
    }

    public boolean isTolerantSchemaEvolution()
    {
        return tolerantSchemaEvolution;
    }
}

File: cpp/pixels-core/lib/reader/PixelsReaderOption.cpp
#include "reader/PixelsReaderOption.h"

PixelsReaderOption::PixelsReaderOption()
{
    skipCorruptRecords = false;
    tolerantSchemaEvolution = true;
    enableEncodedColumnVector = true;
    enableFilterPushDown = false; 
    queryId = -1L;
    batchSize = 0;
    rgStart = 0;
    rgLen = -1;  // -1 means reading to the end of the file
    ringIndex = 0; // 显式初始化
}

void PixelsReaderOption::setIncludeCols(const std::vector <std::string> &columnNames)
{
    includedCols = columnNames;
}

std::vector <std::string> PixelsReaderOption::getIncludedCols()
{
    return includedCols;
}

void PixelsReaderOption::setSkipCorruptRecords(bool s)
{
    skipCorruptRecords = s;
}

bool PixelsReaderOption::isSkipCorruptRecords()
{
    return skipCorruptRecords;
}

void PixelsReaderOption::setQueryId(long qId)
{
    queryId = qId;
}

long PixelsReaderOption::getQueryId()
{
    return queryId;
}

int PixelsReaderOption::getRingIndex() const {
    return ringIndex;
}

void PixelsReaderOption::setRingIndex(int i) {
    ringIndex = i;
}

void PixelsReaderOption::setRGRange(int start, int len)
{
    rgStart = start;
    rgLen = len;
}

int PixelsReaderOption::getRGStart()
{
    return rgStart;
}

int PixelsReaderOption::getRGLen()
{
    return rgLen;
}

void PixelsReaderOption::setTolerantSchemaEvolution(bool t)
{
    tolerantSchemaEvolution = t;
}

bool PixelsReaderOption::isTolerantSchemaEvolution()
{
    return tolerantSchemaEvolution;
}

void PixelsReaderOption::setEnableEncodedColumnVector(bool enabled)
{
    enableEncodedColumnVector = enabled;
}

bool PixelsReaderOption::isEnableEncodedColumnVector()
{
    return enableEncodedColumnVector;
}

void PixelsReaderOption::setEnabledFilterPushDown(bool enabledFilterPushDown)
{
    this->enableFilterPushDown = enabledFilterPushDown;
}

bool PixelsReaderOption::isEnabledFilterPushDown()
{
    return this->enableFilterPushDown;
}


void PixelsReaderOption::setFilter(pixels::TableFilterSet f) {
    //std::cout<<"at set filter"<<f.filters.size()<<std::endl;
    filter = std::move(f);
}

pixels::TableFilterSet PixelsReaderOption::extractFilter() {
    return std::move(filter);
}

int PixelsReaderOption::getfiltersize(){
    return filter.filters.size();
}

void PixelsReaderOption::setBatchSize(int batchSize)
{
    this->batchSize = batchSize;
}

int PixelsReaderOption::getBatchSize() const
{
    return batchSize;
}